### PR TITLE
[API] Add LLaMA decoder API.

### PR DESCRIPTION
**Pull request by changqi1**

### CODE REVIEW

**Comment by changqi1** [comment]
@pujiang2018 Do you have some idea to export layernorm or rms_norm param in the decoder API?
void invokeLayerLLaMA(DataType dt, ActivationType at, NormType nt, ....)

File: include/dtype.h [hunk 1]
```diff
@@ -38,4 +38,17 @@ enum DeviceKind {
     iCPU = 0,
     iGPU,
 };
+
+enum NormType {
+    RMS = 0,
+    Layer,
```
Comment (pujiang2018) [inline include/dtype.h:44]: Is "LN" better than "Layer"?

File: src/layers/decoder_layer.cpp [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+        const float *valueBias = nullptr, const float *attnOutBias = nullptr) {
+    static std::mutex mutex;
+    std::lock_guard<std::mutex> lock(mutex);
+
+    auto prepareAttnMask = [&](DecoderContext *ctx, int step) {
+        int seqLen = ctx->inputSeqLen;
+        int accSeqLen = pastSeqLen + currentSeqLen;
+        float *mask = nullptr;
+
+        auto getAttnMask = [](int sizeRequired) {
```
Comment (pujiang2018) [inline src/layers/decoder_layer.cpp:42]: we are deprecating attention mask, so need to change this in future.

**Comment by pujiang2018** [comment]
> @pujiang2018 Do you have some idea to export layernorm or rms_norm param in the decoder API? void invokeLayerLLaMA(DataType dt, ActivationType at, NormType nt, ....)

I think like this is OK.

File: include/dtype.h (hunk 1 above)
Comment (changqi1) [reply include/dtype.h]: Done

File: src/layers/decoder_layer.cpp (hunk 2 above)
Comment (changqi1) [reply src/layers/decoder_layer.cpp:42]: Add TODO.

**Comment by changqi1** [comment]
@pujiang2018  @Duyi-Wang  Done

**Pull request merged**